### PR TITLE
Paginating public GitHub repos identifies the last page

### PR DESCRIPTION
**Pull request by vrto**
### Problem

Our paging logic thru public repositories in GitHub enterprise doesn't gracefully terminate the pagination on the last page. 

Public repositories are paged using [since](https://docs.github.com/en/rest/reference/repos#list-public-repositories) query parameter, which carries the offset where the pagination ought to resume.

The problem is that GitHub API somewhat surprisingly returns an OK (HTTP 200) response when more items are to be served, but NOT-FOUND (HTTP 404) when the provided `since` value is greater than the last known entry ID.

This error bubbles up to the user interface, which we don't want to.

### Solution

Checking for status code (200 vs. 404) might not be enough, as GitHub can serve a 404 response for many different reasons.

The proper way to determine whether we've reached the last page (hence an empty result) is by looking at the `Link` response header. 

It carries a `first` relationship attribute, and an optional `next` relationship attribute - only when there's another page to look at. If `next` is missing, we know we've reached the end.

b083426f23c2470ea44332e7f7f6b0a0

## Test plan

- see `TestPublicRepos_PaginationTerminatesGracefully` test

### CODE REVIEW

File: internal/repos/github_test.go [hunk 1]
```diff
@@ -107,10 +107,7 @@ func TestGithubSource_GetRepo(t *testing.T) {
 		tc.name = "GITHUB-DOT-COM/" + tc.name
 
 		t.Run(tc.name, func(t *testing.T) {
-			// The GithubSource uses the github.Client under the hood, which
-			// uses rcache, a caching layer that uses Redis.
-			// We need to clear the cache before we run the tests
-			rcache.SetupForTest(t)
+			setUpRcache(t)
```
Comment (vrto) [inline internal/repos/github_test.go:110]: This comment was copy-pasted in the code a few times, so I figured we don't need to repeat ourselves more than once.

File: internal/repos/github_test.go [hunk 2]
```diff
@@ -142,6 +139,57 @@ func TestGithubSource_GetRepo(t *testing.T) {
 	}
 }
 
+func setUpRcache(t *testing.T) {
```
Comment (vrto) [inline internal/repos/github_test.go:142]: This comment was copy-pasted in the code a few times, so I figured we don't need to repeat ourselves more than once.

**Comment by vrto** [comment]
Fixes https://github.com/sourcegraph/sourcegraph/issues/31549

File: internal/extsvc/github/v3.go [hunk 3]
```diff
@@ -701,7 +701,13 @@ func (c *V3Client) ListInstallationRepositories(ctx context.Context, page int) (
 // - /user/repos
 func (c *V3Client) listRepositories(ctx context.Context, requestURI string) ([]*Repository, error) {
 	var restRepos []restRepository
-	if _, err := c.get(ctx, requestURI, &restRepos); err != nil {
+	if res, err := c.get(ctx, requestURI, &restRepos); err != nil {
+		link := res.headers.Get("Link")
+		// if we've reached beyond the last page then GitHub API returns 404 with link to the first page, but does NOT contain link to the next page
+		// link to the next page is typically included in 200 response Link header
```
Comment (ryanslade) [inline internal/extsvc/github/v3.go:707]:
```suggestion
		// If we've reached beyond the last page then GitHub API returns 404 with link to the first page, but does NOT contain link to the next page.
		// link to the next page is typically included in 200 response Link header
```

**Review by ryanslade — APPROVED** [review]
Nice, LGTM!

**Comment by ryanslade** [comment]
@sourcegraph/repo-management FYI

**Pull request merged**